sentry-actix/src/lib.rs

```diff
@@ -119,230 +106,200 @@ impl SentryMiddlewareBuilder {
     }
 }
 
-/// Reports certain failures to sentry.
-pub struct SentryMiddleware {
+/// Reports certain failures to Sentry.
+#[derive(Clone)]
+pub struct Sentry {
     hub: Option<Arc<Hub>>,
     emit_header: bool,
     capture_server_errors: bool,
 }
 
-struct HubWrapper {
-    hub: Arc<Hub>,
-    root_scope: RefCell<Option<ScopeGuard>>,
-}
-
-impl SentryMiddleware {
+impl Sentry {
     /// Creates a new sentry middleware.
-    pub fn new() -> SentryMiddleware {
-        SentryMiddleware {
+    pub fn new() -> Self {
+        Sentry {
             hub: None,
             emit_header: false,
             capture_server_errors: true,
         }
     }
 
     /// Creates a new middleware builder.
-    pub fn builder() -> SentryMiddlewareBuilder {
-        SentryMiddleware::new().into_builder()
+    pub fn builder() -> SentryBuilder {
+        Sentry::new().into_builder()
     }
 
     /// Converts the middleware into a builder.
-    pub fn into_builder(self) -> SentryMiddlewareBuilder {
-        SentryMiddlewareBuilder { middleware: self }
+    pub fn into_builder(self) -> SentryBuilder {
+        SentryBuilder { middleware: self }
     }
+}
 
-    fn new_hub(&self) -> Arc<Hub> {
-        Arc::new(Hub::new_from_top(Hub::main()))
+impl Default for Sentry {
+    fn default() -> Self {
+        Sentry::new()
     }
 }
 
-impl Default for SentryMiddleware {
-    fn default() -> Self {
-        SentryMiddleware::new()
+impl<S, B> Transform<S> for Sentry
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Transform = SentryMiddleware<S>;
+    type InitError = ();
+    type Future = Ready<Result<Self::Transform, Self::InitError>>;
+
+    fn new_transform(&self, service: S) -> Self::Future {
+        ok(SentryMiddleware {
+            service,
+            inner: self.clone(),
+        })
+    }
+}
+
+/// The middleware for individual services.
+pub struct SentryMiddleware<S> {
+    service: S,
+    inner: Sentry,
+}
+
+impl<S, B> Service for SentryMiddleware<S>
+where
+    S: Service<Request = ServiceRequest, Response = ServiceResponse<B>, Error = Error>,
+    S::Future: 'static,
+{
+    type Request = ServiceRequest;
+    type Response = ServiceResponse<B>;
+    type Error = Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>>>>;
+
+    fn poll_ready(
+        &mut self,
+        cx: &mut std::task::Context<'_>,
+    ) -> std::task::Poll<Result<(), Self::Error>> {
+        self.service.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: ServiceRequest) -> Self::Future {
+        let inner = self.inner.clone();
+        let hub = inner
+            .hub
+            .clone()
+            .unwrap_or_else(|| Arc::new(Hub::new_from_top(Hub::main())));
```

Comment:
This should rather be `Hub::new_from_top(inner.hub.unwrap_or_else(Hub::main))`

The same comment about multiple requests pushing and poping scopes from the *same* hub in arbitrary order applies here. Each request has to have its own Hub instance.

Also, I think we don’t need to use `push_scope` at all in that case.